web/src/message_fetch.ts

```diff
@@ -65,24 +97,30 @@ export function load_messages_around_anchor(anchor, cont, msg_list_data) {
     });
 }
 
-function process_result(data, opts) {
-    let messages = data.messages;
+function process_result(data: MessageData, opts: MessageOptions): void {
+    const raw_messages = data.messages;
 
-    messages = messages.map((message) => message_helper.process_new_message(message));
+    const messages = raw_messages.map((raw_message) =>
+        message_helper.process_new_message(raw_message),
+    );
     const has_found_oldest = opts.msg_list?.data.fetch_status.has_found_oldest() ?? false;
     const has_found_newest = opts.msg_list?.data.fetch_status.has_found_newest() ?? false;
 
     // In some rare situations, we expect to discover new unread
     // messages not tracked in unread.ts during this fetching process.
     message_util.do_unread_count_updates(messages, true);
 
+    const msg_list_data = opts.msg_list_data ?? opts.msg_list?.data;
+
+    assert(msg_list_data !== undefined);
```

Comment:
At least one of the two properties must exist, but we cannot be sure which one.
Asserting felt correct to me for this.

Reply:
TypeScript will understand this without an `assert` if you fix the type of `opts`:

```ts
type MessageOptions = {
    anchor: string | number;
    num_before: number;
    num_after: number;
    cont: (data: MessageData, args: MessageOptions) => void;
    fetch_again?: boolean;
} & (
    | {msg_list: message_lists.MessageList; msg_list_data?: MessageListData | undefined}
    | {msg_list?: undefined; msg_list_data: MessageListData}
);
```

Reply:
I've tried this type change, TypeScript is still incapable of understanding that one of them will never be undefined unless we use a discriminated union key.

Reply:
Sorry, I pasted the wrong version. Edited. TypeScript does understand this.